File: ios/Cloudmoji/Cloudmoji/Views/CountTile.swift
import SwiftUI

/// Every number Count mode's grid is drawn from.
///
/// Transcribed from `src/components/CountMode.tsx`. The sizes are step functions
/// of the round size because a round of nine 96pt tiles does not fit on a phone —
/// and the steps stop at 64, because shrinking a child's target below the floor to
/// make a layout fit is the one trade this project does not make.
enum CountTileMetrics {
    /// The floor. Named rather than inlined so the tests can say what they are
    /// enforcing, and so nobody has to guess whether 64 or 72 applies here — a
    /// count tile is child-facing, and 72 is *preferred*, not required.
    static let childMinimum: CGFloat = 64

    static func side(count: Int, compact: Bool) -> CGFloat {
        if compact { return count <= 5 ? 72 : 64 }
        if count <= 3 { return 96 }
        if count <= 6 { return 82 }
        return 72
    }

    static func glyphSize(count: Int, compact: Bool) -> CGFloat {
        if compact { return count <= 5 ? 44 : 36 }
        if count <= 3 { return 64 }
        if count <= 6 { return 54 }
        return 46
    }

    /// Three across upright, five sideways — and never more columns than there are
    /// tiles. The web hard-codes three upright, which leaves a round of two hugging
    /// the left of a three-column track instead of centred under the readout.
    static func columns(count: Int, compact: Bool) -> Int {
        min(count, compact ? 5 : 3)
    }

    /// Roomier for the small rounds, which have the space for it. Never under the
    /// 8pt floor between two things a child taps.
    static func gridSpacing(count: Int, compact: Bool) -> CGFloat {
        if compact { return 10 }
        return count <= 4 ? 16 : 12
    }

    /// `max-width: 360` upright, `520` sideways. The grid centres inside whatever
    /// it is given, so this stops a round of two spreading across an iPad.
    static func maxGridWidth(compact: Bool) -> CGFloat {
        compact ? 520 : 360
    }

    static let cornerRadius: CGFloat = 22
    /// `2.5px solid`. Heavier than a plate hairline because on a counted tile this
    /// border is the state.
    static let borderWidth: CGFloat = 2.5

    /// The order badge: a 34pt disc hung off the top-right corner.
    static let badgeSide: CGFloat = 34
    /// How far it hangs outside the tile, on both axes. The grid pads by this much
    /// so a top-row badge is not clipped.
    static let badgeOverhang: CGFloat = 10
    static let badgeGlyphSize: CGFloat = 19
    static let badgeBorderWidth: CGFloat = 2.5

    /// Uncounted tiles sit back so the counted ones come forward.
    static let uncountedOpacity: Double = 0.75
    static let uncountedScale: CGFloat = 0.95

    /// Design system Active States: this is a tile, so `scale(0.85)` — the same as
    /// an emoji tile, not the 0.88 of a control button.
    static let pressedScale: CGFloat = 0.85

    /// `bounceEmoji 0.35s ease`, fired on the tile just counted.
    static let bounceScale: CGFloat = 1.15
    static let bounceDuration: Double = 0.35
    /// `transition: all 0.25s ease` between the counted and uncounted looks.
    static let stateDuration: Double = 0.25
}

/// One of the N identical things on screen.
///
/// It knows nothing about the round: it is handed a badge number or `nil`, and a
/// size. That is what lets it be measured on its own, and it is why the round's
/// rules live in `CountRound` where they can be tested.
struct CountTile: View {
    let emoji: String
    /// Position in the grid, for the accessibility identifier only. Counting order
    /// is `badge`, and the two are deliberately different: the child counts the
    /// tiles in whatever order they like.
    let index: Int
    /// The number this tile was counted as, or `nil` if it has not been counted.
    let badge: Int?
    /// True for the tile counted most recently, which bounces once.
    let isJustCounted: Bool
    let side: CGFloat
    let glyphSize: CGFloat
    let onTap: () -> Void

    private var isCounted: Bool { badge != nil }

    private var shape: RoundedRectangle {
        RoundedRectangle(cornerRadius: CountTileMetrics.cornerRadius, style: .continuous)
    }

    var body: some View {
        Button(action: onTap) {
            Text(emoji)
                .font(.system(size: glyphSize))
                .frame(width: side, height: side)
                .background(
                    isCounted ? Theme.teal.opacity(0.15) : Theme.surface,
                    in: shape
                )
                .overlay(
                    shape.stroke(
                        isCounted ? Theme.teal.opacity(0.6) : Theme.surfaceBorderStrong,
                        lineWidth: CountTileMetrics.borderWidth
                    )
                )
                // Without an explicit hit shape the tappable area is the glyph's
                // own box and the corners of the square a toddler aims at are dead.
                .contentShape(Rectangle())
                .overlay(alignment: .topTrailing) { badgeView }
        }
        .buttonStyle(PressScale(scale: CountTileMetrics.pressedScale))
        .opacity(isCounted ? 1 : CountTileMetrics.uncountedOpacity)
        .scaleEffect(isJustCounted
                     ? CountTileMetrics.bounceScale
                     : (isCounted ? 1 : CountTileMetrics.uncountedScale))
        // The badge hangs 10pt outside the tile and a bouncing tile overlaps its
        // neighbours, and a grid paints in order — so without this the tile the
        // child just touched has its badge drawn *underneath* the tiles after it.
        // Visible on a device, invisible to every assertion in this suite.
        .zIndex(isCounted || isJustCounted ? 1 : 0)
        .animation(.spring(duration: CountTileMetrics.bounceDuration), value: isJustCounted)
        .animation(.easeInOut(duration: CountTileMetrics.stateDuration), value: isCounted)
        .accessibilityLabel(Text(verbatim: emoji))
        // The badge number, so a UI test can assert *which* number a tile was
        // counted as without reading pixels. Empty when uncounted.
        .accessibilityValue(badge.map(String.init) ?? "")
        .accessibilityIdentifier("count-item-\(index)")
    }

    @ViewBuilder private var badgeView: some View {
        if let badge {
            Text(String(badge))
                // A digit is text, not a colour emoji, so it takes the button's
                // tint — system accent blue — unless this says otherwise.
                .font(Theme.body(CountTileMetrics.badgeGlyphSize, .black))
                .foregroundStyle(Theme.textPrimary)
                .frame(width: CountTileMetrics.badgeSide, height: CountTileMetrics.badgeSide)
                .background(
                    // `linear-gradient(135deg, #4ECDC4, #44B8AC)`.
                    LinearGradient(
                        colors: [Theme.teal, Theme.tealDeep],
                        startPoint: .topLeading,
                        endPoint: .bottomTrailing
                    ),
                    in: Circle()
                )
                .overlay(
                    Circle().stroke(
                        Theme.textPrimary.opacity(0.3),
                        lineWidth: CountTileMetrics.badgeBorderWidth
                    )
                )
                // Flatten before the shadow, or the disc, the ring and the digit
                // each cast their own and the badge reads as three stacked things.
                .compositingGroup()
                .shadow(color: Theme.teal.opacity(0.4), radius: 4, x: 0, y: 2)
                .offset(x: CountTileMetrics.badgeOverhang, y: -CountTileMetrics.badgeOverhang)
                .transition(.scale.combined(with: .opacity))
        }
    }
}

#Preview("Counted and uncounted") {
    ZStack {
        Theme.background.ignoresSafeArea()
        HStack(spacing: 16) {
            CountTile(emoji: "🐶", index: 0, badge: 1, isJustCounted: false,
                      side: 96, glyphSize: 64, onTap: {})
            CountTile(emoji: "🐶", index: 1, badge: nil, isJustCounted: false,
                      side: 96, glyphSize: 64, onTap: {})
            CountTile(emoji: "🐶", index: 2, badge: 12, isJustCounted: true,
                      side: 96, glyphSize: 64, onTap: {})
        }
        .padding(24)
    }
}

File: android/app/src/main/java/app/cloudmoji/android/ui/count/CountTile.kt
package app.cloudmoji.android.ui.count

import androidx.compose.animation.core.animateFloatAsState
import androidx.compose.animation.core.tween
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.interaction.MutableInteractionSource
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.offset
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.remember
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.alpha
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Brush
import androidx.compose.ui.graphics.graphicsLayer
import androidx.compose.ui.platform.testTag
import androidx.compose.ui.semantics.Role
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.role
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.semantics.stateDescription
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.TextUnit
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.compose.ui.zIndex
import app.cloudmoji.android.ui.common.pressScale
import app.cloudmoji.android.ui.theme.CloudmojiBodyFont
import app.cloudmoji.android.ui.theme.Surface
import app.cloudmoji.android.ui.theme.SurfaceBorder
import app.cloudmoji.android.ui.theme.Teal
import app.cloudmoji.android.ui.theme.TealDeep
import app.cloudmoji.android.ui.theme.TextPrimary

/**
 * Every number Count mode's grid is drawn from. Mirrors iOS `CountTile.swift`'s
 * `CountTileMetrics`, dp for pt.
 *
 * The sizes are step functions of the round size because a round of nine
 * 96dp tiles does not fit on a phone — and the steps stop at [childMinimum],
 * because shrinking a child's target below the floor to make a layout fit is
 * the one trade this project does not make.
 */
object CountTileMetrics {
    /** The floor. Named rather than inlined so the tests can say what they
     * are enforcing — a count tile is child-facing, and 72dp is *preferred*,
     * not required. */
    val childMinimum: Dp = 64.dp

    fun side(count: Int, compact: Boolean): Dp = when {
        compact -> if (count <= 5) 72.dp else 64.dp
        count <= 3 -> 96.dp
        count <= 6 -> 82.dp
        else -> 72.dp
    }

    fun glyphSize(count: Int, compact: Boolean): TextUnit = when {
        compact -> if (count <= 5) 44.sp else 36.sp
        count <= 3 -> 64.sp
        count <= 6 -> 54.sp
        else -> 46.sp
    }

    /** Three across upright, five sideways — and never more columns than
     * there are tiles, so a round of two centres under the readout instead
     * of hugging the left of a three-column track. */
    fun columns(count: Int, compact: Boolean): Int = minOf(count, if (compact) 5 else 3)

    /** Roomier for the small rounds, which have the space for it. Never
     * under the 8dp floor between two things a child taps. */
    fun gridSpacing(count: Int, compact: Boolean): Dp = when {
        compact -> 10.dp
        count <= 4 -> 16.dp
        else -> 12.dp
    }

    /** `max-width: 360` upright, `520` sideways — stops a round of two
     * spreading across a tablet. */
    fun maxGridWidth(compact: Boolean): Dp = if (compact) 520.dp else 360.dp

    val cornerRadius: Dp = 22.dp

    /** Heavier than a plate hairline because on a counted tile this border
     * is the state. */
    val borderWidth: Dp = 2.5.dp

    /** The order badge: a 34dp disc hung off the top-end corner. */
    val badgeSide: Dp = 34.dp

    /** How far it hangs outside the tile, on both axes. The grid pads by
     * this much so a top-row badge is not clipped. */
    val badgeOverhang: Dp = 10.dp
    val badgeGlyphSize: TextUnit = 19.sp
    val badgeBorderWidth: Dp = 2.5.dp

    /** Uncounted tiles sit back so the counted ones come forward. */
    const val uncountedOpacity: Float = 0.75f
    const val uncountedScale: Float = 0.95f

    /** Design system Active States: this is a tile, so `scale(0.85)` — the
     * same as an emoji tile, not the 0.88 of a control button. */
    const val pressedScale: Float = 0.85f

    /** `bounceEmoji 0.35s ease`, fired on the tile just counted. */
    const val bounceScale: Float = 1.15f

    /** `transition: all 0.25s ease` between the counted and uncounted looks
     * — also stands in for the bounce's own duration, a reasonable single
     * animation in place of iOS's two independently-keyed ones. */
    const val stateDurationMs: Int = 250
}

/**
 * One of the N identical things on screen. Ported from iOS `CountTile.swift`.
 *
 * It knows nothing about the round: it is handed a badge number or `null`,
 * and a size. That is what lets it be measured on its own, and it is why the
 * round's rules live in `CountRound`, where they can be tested.
 */
@Composable
fun CountTile(
    emoji: String,
    index: Int,
    badge: Int?,
    isJustCounted: Boolean,
    side: Dp,
    glyphSize: TextUnit,
    onTap: () -> Unit,
    modifier: Modifier = Modifier,
) {
    val interactionSource = remember { MutableInteractionSource() }
    val isCounted = badge != null
    val shape = RoundedCornerShape(CountTileMetrics.cornerRadius)

    val scale by animateFloatAsState(
        targetValue = when {
            isJustCounted -> CountTileMetrics.bounceScale
            isCounted -> 1f
            else -> CountTileMetrics.uncountedScale
        },
        animationSpec = tween(durationMillis = CountTileMetrics.stateDurationMs),
        label = "countTileScale",
    )

    Box(
        contentAlignment = Alignment.TopEnd,
        modifier = modifier
            .size(side)
            // Counted/bouncing tiles paint over their neighbours: the badge
            // hangs outside the tile and a bounced tile overlaps the row
            // after it, so without this a grid painting in catalogue order
            // draws the child's own tap underneath what comes after it.
            .zIndex(if (isCounted || isJustCounted) 1f else 0f),
    ) {
        Box(
            contentAlignment = Alignment.Center,
            modifier = Modifier
                .size(side)
                .graphicsLayer { scaleX = scale; scaleY = scale }
                .pressScale(interactionSource, CountTileMetrics.pressedScale)
                .clip(shape)
                .background(if (isCounted) Teal.copy(alpha = 0.15f) else Surface, shape)
                .border(
                    CountTileMetrics.borderWidth,
                    if (isCounted) Teal.copy(alpha = 0.6f) else SurfaceBorder,
                    shape,
                )
                .clickable(interactionSource = interactionSource, indication = null, role = Role.Button, onClick = onTap)
                .alpha(if (isCounted) 1f else CountTileMetrics.uncountedOpacity)
                .semantics {
                    role = Role.Button
                    contentDescription = emoji
                    if (badge != null) stateDescription = "Counted $badge"
                }
                .testTag("count-item-$index"),
        ) {
            Text(text = emoji, fontSize = glyphSize)
        }

        if (badge != null) {
            CountBadge(number = badge)
        }
    }
}

@Composable
private fun CountBadge(number: Int) {
    Box(
        contentAlignment = Alignment.Center,
        modifier = Modifier
            .size(CountTileMetrics.badgeSide)
            .offset(x = CountTileMetrics.badgeOverhang, y = -CountTileMetrics.badgeOverhang)
            .clip(CircleShape)
            .background(Brush.linearGradient(listOf(Teal, TealDeep)), CircleShape)
            .border(CountTileMetrics.badgeBorderWidth, TextPrimary.copy(alpha = 0.3f), CircleShape),
    ) {
        Text(
            text = number.toString(),
            color = TextPrimary,
            fontFamily = CloudmojiBodyFont,
            fontWeight = FontWeight.Black,
            fontSize = CountTileMetrics.badgeGlyphSize,
        )
    }
}
